service/src/main/java/bio/terra/pipelines/common/utils/pagination/PageSpecification.java

```diff
@@ -0,0 +1,63 @@
+package bio.terra.pipelines.common.utils.pagination;
+
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.data.jpa.domain.Specification;
+
+/**
+ * This class is used to help build a predicate to both filter by greater than or less than a field
+ * name and value as well as ordering the results of the query, currently this is hardcoded to
+ * descending to get the most recent records first. This is done based on the CursorBasePageable
+ * passed in.
+ *
+ * @param <T> - DB entity you are filtering on
+ */
+@RequiredArgsConstructor
+public class PageSpecification<T> implements Specification<T> {
+
+  private final transient String mainFieldName;
+  private final transient CursorBasedPageable cursorBasedPageable;
+
+  @Override
+  public Predicate toPredicate(
+      @NotNull Root<T> root, CriteriaQuery<?> query, @NotNull CriteriaBuilder criteriaBuilder) {
+    var predicate = applyPaginationFilter(root, criteriaBuilder);
+    query.orderBy(criteriaBuilder.desc(root.get(mainFieldName)));
```

Comment:
how does this operation on `query` get applied to / affect `predicate`?

Reply:
yea so essentially the idea is all of these predicates get applied in the order you instantiate them. So for us its in the PipelineRunService class with the code

```
    var postSlice =
        pipelineRunsRepository.findAll(
            userIdSpecification.and(pageSpecification), ofSize(pageable.getSize()));
```

so it basically takes the predicate from `userIdSpecification` and  `AND`s it with the `pageSpecification` predicate to build the final criteria out.  Essentially `Predicate`s are used to fill in the `WHERE x AND y AND z` part of a sql query and stuff outside of that like `ORDER BY`, `GROUP BY` can be applied way through the `CriteriaQuery` object